### DISCUSSION TITLE
VCMI port

### DISCUSSION THREAD

**Discussion by Shatur** [eduardofilo/RG350_adam_image#325] [Ports]
I successfully compiled VCMI using the latest OpenDingux toolchain with a few changes which I going to upstream, see https://github.com/vcmi/vcmi/pull/1227.

The game runs:
![IMG_20221213_040231](https://user-images.githubusercontent.com/22453358/207207977-2170b41a-3eb7-4416-8bcc-3ef43ca67df2.jpg)

But I can't start a gaming session and I think it's related to the device configuration. When a scenario or campaign starts, VCMI starts a local server at 127.0.0.1:3030 and tries to connect to it. But for some unknown reason, VCMI fails to connect to it. Log: [gmenu2x.log](https://github.com/vcmi/vcmi/files/10216418/gmenu2x.log).
I tried pinging localhost, but it fails:
```
rg280v:~ $ ping 127.0.0.1
PING 127.0.0.1 (127.0.0.1): 56 data bytes
^C
--- 127.0.0.1 ping statistics ---
39 packets transmitted, 0 packets received, 100% packet loss
```
But I can ping my PC from the device. Also `ip addr` shows the following:
```
rg280v:~ $ ip addr
1: lo: <LOOPBACK> mtu 65536 qdisc noop qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: sit0@NONE: <NOARP> mtu 1480 qdisc noop qlen 1000
    link/sit 0.0.0.0 brd 0.0.0.0
3: usb0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast qlen 1000
    link/ether 0a:8f:8e:1b:a8:45 brd ff:ff:ff:ff:ff:ff
    inet 10.1.1.2/30 scope global usb0
       valid_lft forever preferred_lft forever
    inet6 fe80::88f:8eff:fe1b:a845/64 scope link 
       valid_lft forever preferred_lft forever
```
On my PC `loopback` looks this way:
```
~ $ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
```
@eduardofilo Maybe you know what could be the problem? Maybe loopback is disabled on this firmware or something like that?

**Comment by eduardofilo** [comment]
It seems to be activated, but the classic IP 127.0.0.1 is certainly not associated with it.
You could try changing the loopback connections to 10.1.1.2, although I suppose that for this interface to exist you will have to connect the USB cable to the computer. I don't know, don't think I have too much idea.

**Reply by Shatur** [reply]
> It seems to be activated, but the classic IP 127.0.0.1 is certainly not associated with it.

Makes sense... I tried calling `ip addr add 127.0.0.1/8 dev lo` and now 127.0.0.1 appears in `ip addr`, but I still can't connect to it. Maybe something else needs to be done, will investigate. Also I noticed that I can't ssh to 127.0.0.1. Doesn't make much sense, but it's usually possible on a machine with ssh server running.

> You could try changing the loopback connections to 10.1.1.2, although I suppose that for this interface to exist you will have to connect the USB cable to the computer. I don't know, don't think I have too much idea.

I probably could try to hack the game  to see if it is even playable, yes.
Thanks for looking into it!

**Reply by Shatur** [reply]
A friend helped me figure this out! It turns out loopback was actually disabled:
```
rg280v:~ $  ip link show dev lo
1: lo: <LOOPBACK> mtu 65536 qdisc noop qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
```
And can be enabled using `ip link set lo up`. After it `127.0.0.1/8` and `::1/128` are displayed in `ip addr`. The previously mentioned command `ip addr add 127.0.0.1/8 dev lo` is unnecessary.

**Comment by Shatur** [comment]
It works! But unplayable because the UI is not adaptive. I have to run in 2.5x resolution in order to fit everything in the screen. But elements and text becomes very small. Also in game the playing field is displayed at the top and takes only half of the screen. Bottom consists of half of a half of a loading screen and pieces of old dialogues (just a section that is not redrawn). And performance is very low. I think that this is also due to the fact that the game runs at 2.5x resolution.
Maybe the game will be more adaptive in the future.
![IMG_20221213_220143](https://user-images.githubusercontent.com/22453358/207433539-3104da7b-e285-4cfc-bb4f-8ce061414c76.jpg)

**Reply by eduardofilo** [reply]
Good try.
